types/aws-lambda/index.d.ts

```diff
@@ -569,6 +570,30 @@ export interface CloudFrontHeaders {
     }>;
 }
 
+export interface CloudFrontOrigin {
+    custom?: CloudFrontCustomOrigin;
+    s3?: CloudFrontS3Origin;
+}
```

Comment:
Since it's declared you must have exactly one of these, a stronger definition would be:

```ts
// the never fields are required both to require only one field is set and
// to allow reading if either field exists.
type CloudFrontOrigin =
    | { s3: CloudFrontS3Origin, custom?: never }
    | { custom: CloudFrontCustomOrigin, s3?: never };
```

but I'm not sure it's worth the extra complexity. If more fields are added in the future, this type gets ugly real fast.

Reply:
Yes, that solves the problem of the mutually exclusive options but doesn't help with the reuse of the request object in the response, which does not carry the origin field.

I like the idea of enforcing the mutual exclusivity, so let me know if you don't want this added.

Reply:
If you have code that uses this and it works, sure, go ahead!

The docs seem to say you can modify the origin on inbound requests, though?

> In the CloudFront origin request event, the origin element in the event structure contains information about the Amazon S3 or custom origin that the request would be routed to, based on the path pattern. You can update the values in the origin object to route a request to a different origin. The origin doesn't need to be defined in the distribution, and you can replace an Amazon S3 origin object with a custom origin object and vice versa.

I suppose you could `interface CloudFrontInboundRequest extends CloudFrontRequest { origin?: CloudFrontOrigin; }`, but that's not just more complex a definition, but adding more types for the user to think about and has back-compatibility issues - I'd say that one isn't worth it.